src/src/PluginStructs/P104_data_struct.cpp

```diff
@@ -0,0 +1,2312 @@
+      displayOneZoneText(zstruct.zone - 1, zstruct, sZoneInitial[zstruct.zone - 1]); // Re-send last displayed text
+      P->displayReset(zstruct.zone - 1);
+    }
+    # ifdef P104_USE_BAR_GRAPH
+
+    if (zstruct.content == P104_CONTENT_BAR_GRAPH) {
+      displayBarGraph(zstruct.zone - 1, zstruct, sZoneInitial[zstruct.zone - 1]); // Re-send last displayed bar graph
+    }
+    # endif // ifdef P104_USE_BAR_GRAPH
+
+    // Repeat timer is/should be started elsewhere
+  }
+}
+
+# ifdef P104_USE_BAR_GRAPH
+
+/***********************************************
+ * Enable/Disable updating a range of modules
+ **********************************************/
+void P104_data_struct::modulesOnOff(uint8_t start, uint8_t end, MD_MAX72XX::controlValue_t on_off) {
+  for (uint8_t m = start; m <= end; m++) {
+    pM->control(m, MD_MAX72XX::UPDATE, on_off);
+  }
+}
+
+/********************************************************
+ * draw a single bar-graph, arguments already adjusted for direction
+ *******************************************************/
+void P104_data_struct::drawOneBarGraph(uint16_t lower,
+                                       uint16_t upper,
+                                       int16_t  pixBottom,
+                                       int16_t  pixTop,
+                                       uint16_t zeroPoint,
+                                       uint8_t  barWidth,
+                                       uint8_t  barType,
+                                       uint8_t  row) {
+  bool on_off;
+
+  for (uint8_t r = 0; r < barWidth; r++) {
+    for (uint8_t col = lower; col <= upper; col++) {
+      on_off = (col >= pixBottom && col <= pixTop); // valid area
+
+      if ((zeroPoint != 0) &&
+          (barType == P104_BARTYPE_STANDARD) &&
+          (barWidth > 2) &&
+          ((r == 0) || (r == barWidth - 1)) &&
+          (col == lower + zeroPoint)) {
+        on_off = false; // when bar wider than 2, turn off zeropoint top and bottom led
+      }
+
+      if ((barType == P104_BARTYPE_SINGLE) && (r > 0)) {
+        on_off = false; // barType 1 = only a single line is drawn, independent of the width
+      }
+
+      if ((barType == P104_BARTYPE_ALT_DOT) && (barWidth > 1) && on_off) {
+        on_off = ((r % 2) == (col % 2)); // barType 2 = dotted line when bar is wider than 1 pixel
+      }
+      pM->setPoint(row + r, col, on_off);
+    }
+  }
+}
+
+/********************************************************************
+ * Process a graph-string to display in a zone, format:
+ * value,max-value,min-value,direction,bartype|...
+ *******************************************************************/
+void P104_data_struct::displayBarGraph(uint8_t                 zone,
+                                       const P104_zone_struct& zstruct,
+                                       const String          & graph) {
+  if (graph.isEmpty()) { return; }
```

Comment:
This `displayBarGraph` function does quite a lot, including lots of for/while loops.
It also does communicate it to the display driver.
And to make matters worse, you call it from loops too.

So maybe you should add `delay(0)` at the end of this function, to make sure you're not running into other issues where the IP stack receives messages which are just buffered and not processed.
Those will also add up to the memory usage.

Reply:
It is only called for displaying a bar graph, and the loops should be relatively compact/short as there are max. 8 bars to be displayed.
I've added a `delay(0)` in the actual drawing function, and a few more in the parsing and processing loops.